### PR TITLE
[swift2objc] feat:support swift tuples(return types)

### LINKED ISSUE #1783: [swift2objc] Support tuples
Like Dart records, Swift tuples can have positional and named elements. The main difference is that to access the first element of a Dart record you write `foo.$1`, whereas in Swift you write `foo.0`.

Tuples are easy enough to parse, but they're not supported in `@objc` annotated APIs. So we'll need to transform them into classes.

```Swift
(Int, String, abc: MyClass)

// Would be transformed to something like:

public class Tuple_Int_String_abc_MyClass {
    public let _0: Int
    public let _1: String
    public let abc: MyClassWrapper
}
```

### PR DESCRIPTION
**Pull request by Hassnaa9**
This PR implements support for Swift tuples as return types in swift2objc, they are transformed into generated Objective-C compatible wrapper classes.

Closes #1783

### CODE REVIEW

File: pkgs/swift2objc/test/integration/tuples_input.swift [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Test simple tuple return
+public class TupleTest {
```
Comment (liamappelbe) [inline pkgs/swift2objc/test/integration/tuples_input.swift:6]: Cool. Can you also add cases for 0 element tuples, `()`, and 1 element tuples, `(Int)`/`(value: Int)`. I know 0 element tuples are the same as `Void`, and 1 element tuples are just unwrapped to the 1 type, but I want to make sure that swift2objc can handle these cases gracefully.

File: pkgs/swift2objc/test/integration/tuples_output.swift [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    @objc public func getOptionalTuple() -> Tuple_Int_String? {
+      let result = wrappedInstance.getOptionalTuple()
+      return result == nil ? nil : Tuple_Int_String(_0: result!.0, _1: result!.1)
+    }
+
+  }
+
+}
+
+@objc public class Tuple_Bool_Double {
```
Comment (liamappelbe) [inline pkgs/swift2objc/test/integration/tuples_output.swift]: This class will need to inherit from `NSObject`.

File: pkgs/swift2objc/test/integration/tuples_output.swift [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+  public let _0: Bool
```
Comment (liamappelbe) [inline pkgs/swift2objc/test/integration/tuples_output.swift]: The pattern we've used everywhere else is to have the wrapper class hold an instance of the wrapped class, called `wrappedInstance`. So in this case you'd have a private field `var (Bool, Double) wrappedInstance`. I'd like to follow that pattern here if possible. Then you have getters/setters that simply forward to the corresponding getter/setter of the `wrappedInstance`. Should simplify your codegen a bit.

File: pkgs/swift2objc/lib/src/transformer/transformers/transform_referred_type.dart [hunk 4]
```diff
@@ -32,3 +38,122 @@ ReferredType transformReferredType(
...
+
+String _sanitizeTypeName(ReferredType type) {
+  return type.swiftType
+      .replaceAll('<', '_')
+      .replaceAll('>', '')
+      .replaceAll(',', '')
+      .replaceAll(' ', '')
+      .replaceAll('?', 'Optional')
+      .replaceAll('[', 'Array_')
+      .replaceAll(']', '');
```
Comment (liamappelbe) [inline pkgs/swift2objc/lib/src/transformer/transformers/transform_referred_type.dart]: Maybe just hard code `<`, `?`, `[`, and use a regex to catch everything else that maps to `''`. Otherwise it might turn out that you just missed a character, or this might break in future as we add more features. I can already think of things that this misses.

File: pkgs/swift2objc/lib/src/transformer/transform.dart [hunk 5]
```diff
@@ -31,6 +31,26 @@ class TransformationState {
 
   // Bindings that will be generated as stubs.
   final stubs = <Declaration>{};
+
+  // Map from tuple signature to generated wrapper class
+  final _tupleWrappers = <String, ClassDeclaration>{};
+
+  /// Check if a tuple wrapper class has already been generated
+  bool hasGeneratedTuple(String className) =>
```
Comment (liamappelbe) [inline pkgs/swift2objc/lib/src/transformer/transform.dart]: Can you remove these methods and just use `tupleWrappers` directly? They're all just thin wrappers around ordinary map functionality, so there's not much point to them.

File: pkgs/swift2objc/lib/src/transformer/transform.dart [hunk 6]
```diff
@@ -31,6 +31,26 @@ class TransformationState {
...
+      _tupleWrappers.containsKey(className);
+
+  /// Get the wrapper class for a tuple
+  ClassDeclaration getTupleWrapper(String className) =>
+      _tupleWrappers[className]!;
+
+  /// Register a newly generated tuple wrapper class
+  void registerTupleWrapper(String className, ClassDeclaration wrapperClass) {
+    _tupleWrappers[className] = wrapperClass;
+    bindings.add(wrapperClass);
```
Comment (liamappelbe) [inline pkgs/swift2objc/lib/src/transformer/transform.dart]: Manually adding to the bindings like this shouldn't be necessary, and is inconsistent with the design of the transformer. Instead you should probably be reporting the tuple in `ListDeclsVisitation`, or maybe not add them to the bindings at all since you're explicitly adding them to the return list in `transform`.

I also think adding it here is not going to work like you expect. I'm pretty sure `registerTupleWrapper` is only called during `maybeTransformDeclaration`, which is pretty late in the transformation pipeline. For example, by this time, the `globalNamer` has already been constructed, so the tuples won't be including in renaming.

Makes me wonder if maybe a better approach would be for tuples to subclass `Declaration` rather than `ReferredType`. I'm not sure if that would actually be easier or not, since it would probably complicate parsing, as the type parser can't currently create `Declaration`s.

File: pkgs/swift2objc/lib/src/parser/parsers/parse_type.dart [hunk 7]
```diff
@@ -129,11 +129,51 @@ typedef PrefixParselet =
   Json token,
   TokenList fragments,
 ) {
-  final nextToken = fragments[0];
-  if (_tokenId(nextToken) != 'text: )') {
-    throw Exception('Tuples not supported yet, at ${token.path}');
+  if (fragments.isNotEmpty && _tokenId(fragments[0]) == 'text: )') {
```
Comment (liamappelbe) [inline pkgs/swift2objc/lib/src/parser/parsers/parse_type.dart]: Would it be possible to remove this special case and instead check if `elements` is empty below?

File: pkgs/swift2objc/test/integration/tuples_input.swift (hunk 1 above)
Comment (Hassnaa9) [reply pkgs/swift2objc/test/integration/tuples_input.swift:6]: > (value: Int)

This feature isn’t supported in swift2objc.
Swift normalizes () to Void and (Int) to Int (and doesn’t allow (value: Int)), so we don’t treat them as distinct tuple types or generate tuple wrappers for them.

File: pkgs/swift2objc/test/integration/tuples_output.swift [hunk 8]
```diff
@@ -65,199 +73,319 @@ import Foundation
...
 
 }
 
-@objc public class Tuple_Bool_Double {
-  public let _0: Bool
+@objc public class Tuple_Bool_Double: NSObject {
+  var wrappedInstance: (Bool, Double)
+
+  @objc public var _0: Bool {
+    get {
```
Comment (liamappelbe) [inline pkgs/swift2objc/test/integration/tuples_output.swift:108]: Should these have setters too?

**Review by liamappelbe — COMMENTED** [review]
Take a look at the failing test. The issue is that optional primitives aren't supported yet. See https://github.com/dart-lang/native/issues/1743. You can just comment out that case, and leave a TODO linking to that bug. Something like:
`// TODO(https://github.com/dart-lang/native/issues/1743): Enable this`

btw I'm going on vacation for a week, so no more code reviews until next week.

File: pkgs/swift2objc/test/integration/tuples_output.swift (hunk 8 above)
Comment (Hassnaa9) [reply pkgs/swift2objc/test/integration/tuples_output.swift:108]: These wrappers are currently only used as return values, so there's nothing meaningful to set back to.

**Comment by Hassnaa9** [comment]
> btw I'm going on vacation for a week, so no more code reviews until next week.

@liamappelbe Hope you have a relaxing week.
I’ll also spend time this week exploring FFI/ffigen a bit more.

File: pkgs/swift2objc/test/integration/tuples_output.swift (hunk 8 above)
Comment (liamappelbe) [reply pkgs/swift2objc/test/integration/tuples_output.swift:108]: We will probably want to use these for more things in future, so it's best to include setters unless there's a good reason not to. For example, if you have an API accepting a function that returns multiple values, then it'll be important to be able to construct/manipulate tuples when writing that callback.

Comment (Hassnaa9) [reply pkgs/swift2objc/test/integration/tuples_output.swift:108]: > then it'll be important to be able to construct/manipulate tuples when writing that callback.

ooh, I had previously learned that tuples in swift are immutable.
but after make simple search now i know that `Tuples in Swift are mutable if declared with var, but immutable if declared with let`
so i will push the updated code now

File: pkgs/swift2objc/test/integration/tuples_input.swift [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        return (1, 2, 3, 4, 5)
+    }
+    
+    // All labeled elements
+    public func getAllLabeledTuple() -> (x: Int, y: Int, z: String) {
+        return (x: 10, y: 20, z: "point")
+    }
+    
+    public class NestedTupleTest {
+    public func getNestedTuple() -> (Int, (String, Bool)) {
```
Comment (liamappelbe) [inline pkgs/swift2objc/test/integration/tuples_input.swift]: nit: indentation

File: pkgs/swift2objc/test/integration/tuples_input.swift [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    
+    // Tuple with optional elements
+    // TODO(https://github.com/dart-lang/native/issues/1743): Enable this when optional primitives are supported.
+    // public func getTupleWithOptionals() -> (Int?, String?) {
+    //     return (nil, "test")
+    // }
+    
+    // Deeply nested tuple (3 levels)
+    public func getDeeplyNestedTuple() -> (Int, (String, (Bool, Double))) {
```
Comment (liamappelbe) [inline pkgs/swift2objc/test/integration/tuples_input.swift:35]: I thought of another edge case: `((Int, String), (Int, String))`. I think a repeated nested tuple like this will break your current `state.tupleWrappers` caching logic.

File: pkgs/swift2objc/lib/src/generator/generators/class_generator.dart [hunk 11]
```diff
@@ -154,11 +154,20 @@ List<String> _generateClassProperty(PropertyDeclaration property) {
   if (property.isStatic) {
     header.write('static ');
   }
+
   final prefixes = [if (property.unowned) 'unowned', if (property.weak) 'weak'];
 
   var prefix = prefixes.isEmpty ? '' : '${prefixes.join(' ')} ';
   var propSwiftType = property.type.swiftType;
 
+  if (property.getter == null && property.setter == null) {
```
Comment (liamappelbe) [inline pkgs/swift2objc/lib/src/generator/generators/class_generator.dart]: is this special case still needed, now that you've changed the tuple code gen to use the `wrappedInstance` pattern?

File: pkgs/swift2objc/test/integration/tuples_input.swift (hunk 10 above)
Comment (Hassnaa9) [reply pkgs/swift2objc/test/integration/tuples_input.swift:35]: @liamappelbe The caching is functioning correctly after i added this test case and generated the output

File: pkgs/swift2objc/lib/src/generator/generators/class_generator.dart (hunk 11 above)
Comment (Hassnaa9) [reply pkgs/swift2objc/lib/src/generator/generators/class_generator.dart]: you are right, I forgot to refactor it

File: pkgs/swift2objc/test/integration/tuples_output.swift (hunk 8 above)
Comment (liamappelbe) [reply pkgs/swift2objc/test/integration/tuples_output.swift:108]: Ok, I'll have to come up with a trickier example. The bug I'm trying to hit is that you're doing the entire tuple transformation before inserting the tuple into the cache. That transformation will transform all the elements of the tuple too, and is recursive and potentially reentrant. So you could get into a situation where you try to transform a tuple, and in doing so you try to do another transform of that tuple before the first is complete. Since the cache hasn't been written to for the first transform, you start another transform from scratch. Best case you end up with multiple copies of the tuple wrapper class. Worst case you end up in an infinite loop.

I think this example should hit it:

```Swift
class Foo {
    var cycle: (Foo, Bar)?
}

class Bar {
    var cycle: (Foo, Bar)?
}
```

We run into this problem all over the place in all these tools. The fix is to:

1. Check the cache
2. If there's no cache entry for the element, create an empty one (in this case you'd create the wrapper class but not fill it yet), and add it to the cache.
3. Fill out the newly created element.

Step 3 is where the recursive cycle can happen, but by that time the cache already has an entry for this element, so the reentrant calls will return the empty element (which will be filled in later).

Try creating that `Foo`/`Bar` cycle example and see if it causes problems (either infinite loops, or multiple tuples). If it does, apply the fix steps I explained above. If the example doesn't cause any problems, we'll have to try to come up with an even trickier example.

**Comment by liamappelbe** [comment]
Also, please sync to head. Looks like you've got a merge conflict

File: pkgs/swift2objc/test/integration/tuples_output.swift (hunk 8 above)
Comment (Hassnaa9) [reply pkgs/swift2objc/test/integration/tuples_output.swift:108]: @liamappelbe  Good catch, the buggy code generates two separate wrapper classes (`Tuple_TupleTestFoo_TupleTestBar` and `Tuple_TupleTestFoo_TupleTestBar1`).I applied your suggested fix (same pattern as `transformCompound` i think)
now it correctly generates a single shared wrapper class.

**Comment by Hassnaa9** [comment]
@liamappelbe  I created a new PR for the support-swift-tuples branch. I did a rebase instead of a merge,which caused a temporary interruption in the branch history i think
sorry for that :)
new PR: https://github.com/dart-lang/native/pull/3158

**Comment by liamappelbe** [comment]
Ok. Try to figure out what git command you did wrong and make sure it doesn't happen again. This churn makes code reviews harder, and I'm kinda overloaded with code reviews at the moment. One trick is that if your local branch gets messed up, you can delete it and then fetch this PR's branch, as an easy way to restore your local branch to a previous good state.
